members only

vidhishweta01 opened the pull request:
This project is about authentications of user signed in,and the part they access on post i.e. they can access author name when they are signed in, if not register they can only see posts but not the authors name, project is built with Rails.
  here I have used active storage for the avatar attached to the user, and devise is used for authentications.

enaburekhan reviewed (commented):
# Status: Invalid
**Friendly reminder: You should make sure that your repo looks professional before review.
Please take a step back and read a lesson [here](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243642/).
Then apply 3 simple rules from this lesson (it will be easy as you can use a template provided there).
This Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready!
Also, make sure the PR is passing linter checks.**

Happy coding!
Eric Enaburekhan

GzimAsani reviewed (changes requested):
# STATUS: Changes Required :red_circle: 

### Hello @vidhishweta01 good job so far
To highlight: 
- I like the layouts of your app

To improve:
- [ ] Mention in the readme that you must have ImageMagick or GraphicsMagick installed
- [ ] Add back button in posts index page so the user can use go back and forth while using the buttons
- [ ] The app should have only the following routes for posts ->  :posts, only: [:new, :create, :index] ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/posts_routes_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication))
- [ ] You should use `before_action` to restrict access to the `:new` and `:create` methods (uses ApplicationController helper method) ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/authentication/before_action_requirement.png) from [Odin](https://www.theodinproject.com/courses/ruby-on-rails/lessons/authentication)).
- [ ] Remember to open to the pull request from your `feature branch` to your `Development` branch not your `master` ([Git Flow guidelines](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/59882084/))


_Please implement all the above mentioned requirements and ask for another code review whenver you are ready_

### **If you have any question related to the requirements please add a comment here and ping me on slack @GzimAsani**

HAPPY CODING :computer: :desktop_computer:

vidhishweta01 commented:
1. Mentioned in the readme about ImageMagick  installed
2.  Added all post  button in posts index page so the user can use go back and forth while using the buttons
3.  used before_action to restrict access to the :new and :create methods in post_controller
4. added more description as well as screenshots to readme
5. Git flow is being used as devise is the featured branch from master and master is the development branch from main,and  main  branch is not taken as default branch.

MahmoudBakr23 commented on app/views/pages/_header.html.erb:
```diff
@@ -0,0 +1,48 @@
+
+  <div class="collapse navbar-collapse" id="navbarSupportedContent">
+    <ul class="navbar-nav mr-auto">
+    <%if user_signed_in?%>
+      <li class="nav-item">
+      <%= link_to 'Sign out', destroy_user_session_path, method: :delete, class:"nav-link text-light" %>
+      </li>
+      <li>
+      <%= link_to 'Edit Profile', edit_user_registration_path, class:"nav-link text-light"%>
+      </li>
+      <li>
+       <%= link_to 'New Post', new_post_path, class:"nav-link text-light"%>
+       </li>
+    <% else %>
+       <li>
+      <%= link_to 'Sign Up', new_user_registration_path, class:"nav-link text-light"%>
+      </li>
+      <li>
+      <%= link_to 'Sign in', new_user_session_path, class:"nav-link text-light"%>
+      </li>
+      
+    <%end%>
```
Refactor this logic code here!

MahmoudBakr23 commented on app/views/pages/home.html.erb:
```diff
@@ -0,0 +1,29 @@
+<%if current_user%>
+<%#= link_to 'New Post', new_post_path%>
+<div class="float-right rounded">
+<%= image_tag(current_user.avatar_thumbnail, class: "rounded-circle")  %>
+</div>
+<%end%>
```
Refactor this logic code here!

MahmoudBakr23 commented on app/views/posts/index.html.erb:
```diff
@@ -0,0 +1,29 @@
+<p id="notice"><%= notice %></p>
+<%if current_user%>
+<%= link_to 'New Post', new_post_path%>
+<div class="float-right rounded">
+<%= image_tag(current_user.avatar_thumbnail, class: "rounded-circle")  %>
+</div>
+<%end%>
```
Refactor this logic code here!

MahmoudBakr23 commented on app/views/posts/index.html.erb:
```diff
@@ -0,0 +1,29 @@
+        <%= post.blog.truncate(120) %>
+      </div>
+      <%= link_to 'Read More', post_path(post), class: "btn btn-info" %>
+      <%if signed_in? && post.user_id == current_user.id%>
+        <%= link_to 'Edit', edit_post_path(post), class:'btn btn-info text-center mx-1'%>
+        <%= link_to 'Destroy', post, method: :delete, data: { confirm: 'Are you sure?' }, class:'btn btn-info text-center' %>
+      <%end%>
```
Refactor this logic code here!

MahmoudBakr23 commented on app/views/posts/show.html.erb:
```diff
@@ -0,0 +1,24 @@
+<p id="notice"><%= notice %></p>
+<%if current_user%>
+<div class="float-right rounded">
+<%= image_tag(current_user.avatar_thumbnail, class: "rounded-circle")  %>
+</div>
+<%end%>
```
Refactor this logic code here!

MahmoudBakr23 commented on app/views/posts/show.html.erb:
```diff
@@ -0,0 +1,24 @@
+  <%= @post.blog %>
+</p>
+</div>
+<%# <%if current_user && post.user_id == current_user.id%> 
+<%#= link_to 'Edit', edit_post_path(@post) %>
+<%# <%end%> 
+<%#= link_to 'Back', root_path %>
```
Remove this commented-out code.

rukundoeric replied:
Remove this comment

MahmoudBakr23 reviewed (changes requested):
# **Changes Required 🛑♻**

<hr>

Hello there :)
Thanks for all the efforts you made to make the requested changes. However, there are still a few more to pass this. ✔👍

<hr>

- [ ] Kindly move any logical code from the views into the helpers. There should be no logic code in the views.
- [ ] Please remove any commented-out code since you don't use it anyway, and for keeping the code clean.

<hr>

Please do the changes, then ask for another CRR.
Good luck and happy coding 👨‍💻💻

vidhishweta01 commented:
Added method to post_helper
![Screenshot from 2021-04-05 10-40-23](https://user-images.githubusercontent.com/73999611/113539582-b58a4600-95fb-11eb-90f3-eaa9d8cbadc9.png)

rukundoeric commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -0,0 +1,36 @@
+  <%= render "devise/shared/error_messages", resource: resource %>
+<div class="container-fluid w-50 py-2" style="background-color:rgb(255, 219, 175); opacity: 0.84;" >
+  <div class="field">
+  <%if resource.avatar.attached?%>
+  <%=image_tag(resource.avatar)%>
+  <%else%>
+   <%= image_tag('/default_profile.jpg')%>
+  <%end%>
```
move this logic into helper

rukundoeric reviewed (changes requested):
# Status: ♻️ CHANGES REQUIRED ♻️
_Great job guys so far 👏 💯 👍. However, there are few issues we need to address before we move on, please check the comments below._

- ### To Highlight
  - [x] The code looks clean and professional. 💯
  - [x] The project follows Git flow. 💯

- ### To Improve
  - [ ] [MANDATORY] Remove all commented-out code.
  - [ ] [MANDATORY] All logic related code should be moved into helpers. Kindly refer to the previous review. and checkout this [Article](https://www.nopio.com/blog/view-objects-rails-views/) for more information.

_Please make all the required changes before you submit another code review. If you have any problems or concerns, you can reach out to me on Slack: Rukundo Eric 🚀 . I wish you all the very best 🥇_

### _Keep working hard._
### _Happy Coding😃_
-----------------------------------------------------------------------------------------------------------------------------------------
**Regards,
[Rukundo Eric](https://rukundo-eric.herokuapp.com/) ,
[LinkedIn](https://www.linkedin.com/in/rukundo-eric/), 
Slack: Rukundo Eric**

vidhishweta01 commented:
Removed all commented-out code.
 All logic related code is moved into post_helpers.

descholar-ceo reviewed (changes requested):
# Status: Changes Required :vertical_traffic_light: :warning: 
## Hello @vidhishweta01  :handshake: :handshake: :handshake: 

:+1: This is good to make it this far, your project it is really nice, congratulations on the good work you have done so far.
:+1: I liked that you have tried to implement the requested changes from the previous review, keep it up, and your project is really almost done.
#### :warning: However we need to make few changes to mark this project approved, don't feel discouraged because of being asked some changes more and more, that's beacuse this is a learning process, and we are here to help you to become an out-standing developer. See the requested change below :point_down: :point_down: :point_down: 
### Note: All of requested changes which are marked as [Required] are mandatory to be implemented in order to get this project approved, others which are marked as [Optional] are nice to have in your project to make it outstanding

# [Required Changes] :red_circle: 
- [ ] You have violated git/github flow, this pull request should be referring to a development branch instead of master branch, please fix that issue
- [ ] After creating a post, your app is crashing with the following error
![Screenshot from 2021-04-05 09-45-23](https://user-images.githubusercontent.com/56886064/113550836-05650f00-95f4-11eb-9bee-d8e393d2346f.png)

___
_If anything seems to be unclear to you, reachout and ask anything by commenting on this Pull Request and mention me as [descholar-ceo](https://github.com/descholar-ceo), I will receive a notification and will respond._

[GitHub](https://github.com/descholar-ceo) | [LinkedIn](https://linkedin.com/in/mugirase-emmanuel) | [Twitter](https://twitter.com/descholar3)

_Reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)_

vidhishweta01 commented:
Git flow is being used as devise is the featured branch from master and master is the development branch from main,and main branch is not taken as default branch.
that error has been resolved

Its checked before
![Screenshot from 2021-04-05 13-32-18](https://user-images.githubusercontent.com/73999611/113552193-69e39680-9613-11eb-9fde-247e19347884.png)

descholar-ceo reviewed (approved):
# Status: Approved :heavy_check_mark: :heavy_check_mark: :heavy_check_mark:  
## :clap: :clap: :clap: Congratulations @vidhishweta01  , your project is approved :tada: :tada: :tada: 
You have implemented all minimum requirements for this project to mark it as complete, it doesn't need more changes, you can go ahead and merge all of the changes on base branch,

Don't forget to fill the mandatory project completion form, and see you to next project,

Happy coding :computer: :computer: :computer: 

Re-reviewed by [Mugirase Emmanuel (descholar)](https://github.com/descholar-ceo)
